src/imars3dv2/filters/crop.py

```diff
@@ -0,0 +1,162 @@
+    rel_intensity_threshold_air_or_slit: float = 0.05,
+    rel_intensity_threshold_fov: float = 0.1,
+    rel_intensity_threshold_sample: float = 0.95,
+) -> np.ndarray:
+    """
+    Crop the image stack to provided limits or auto detected bound box.
+
+    Case 1: slits in, i.e I_inner >> I_outer
+    o-------------------------------------------------------
+    |        low                                           |
+    |     +++++++++++++++++top slit+++++++++++++++         |
+    |     +                                      +         |
+    |     +                                      +         |
+    |   left slit           FOV(high)         right slit   |
+    |     +                                      +         |
+    |     +                                      +         |
+    |     +++++++++++++++++bottom slit+++++++++++++        |
+    |                                                      |
+    --------------------------------------------------------
+
+    Case 2: slits out, i.e I_inner < I_outer
+    o-------------------------------------------------------
+    |                                                      |
+    |            *****************                         |
+    |             *    object   *           air            |
+    |              *   (low)   *           (high)          |
+    |               ***********                            |
+    |------------------------------------------------------|
+
+    Parameters
+    -----------
+    @param arrays: np.ndarray
+        The image stack to crop. Can also be a 2D image.
+    @param crop_limit: tuple
+        The four limits for cropping. Default is (-1, -1, -1, -1), which will trigger
+        the automatic bounds detection.
+    @param border_pix: int
+        the width of border region to estimate the background intensity, which helps
+        to determine which case we are in.
+    @param expand_ratio: float
+        the ratio to expand the cropped region.
+    @param rel_intensity_threshold_air_or_slit: float
+        passing through keyword arguments to detect_bounds.
+    @param rel_intensity_threshold_fov: float
+        passing through keyword arguments to detect_bounds.
+    @param rel_intensity_threshold_sample: float
+        passing through keyword arguments to detect_bounds.
+
+    Returns
+    -------
+    @return: np.ndarray
+        The cropped image stack.
+    """
+    # check if crop limits (bounding box) are provided
+    if -1 in crop_limit:
+        crop_limit = detect_bounds(
+            arrays,
+            border_pix,
+            expand_ratio,
+            rel_intensity_threshold_air_or_slit,
+            rel_intensity_threshold_fov,
+            rel_intensity_threshold_sample,
+        )
+    # crop
+    left, right, top, bottom = crop_limit
+    if arrays.ndim == 2:
+        return arrays[top:bottom, left:right]
+    elif arrays.ndim == 3:
+        return arrays[:, top:bottom, left:right]
+    else:
```

Comment:
Would it be better to move the check for `arrays.ndim` not being supported to higher up in the function to fail early?

Reply:
I see, I will rearrange the order now